dvc/utils/studio.py

```diff
@@ -0,0 +1,93 @@
+import logging
+import os
+from typing import TYPE_CHECKING, Any, Callable, Dict, List, Optional, Tuple
+from urllib.parse import urljoin
+
+from dvc_studio_client.post_live_metrics import get_studio_repo_url
+from funcy import compact
+from requests import RequestException, Session
+from requests.adapters import HTTPAdapter
+
+if TYPE_CHECKING:
+    from requests import Response
+
+logger = logging.getLogger(__name__)
+
+STUDIO_URL = "https://studio.iterative.ai"
+STUDIO_REPO_URL = "STUDIO_REPO_URL"
+STUDIO_TOKEN = "STUDIO_TOKEN"  # noqa: S105
+
+
+def get_studio_token_and_repo_url(
+    default_token: Optional[str] = None,
+    repo_url_finder: Optional[Callable[[], Optional[str]]] = None,
+) -> Tuple[Optional[str], Optional[str]]:
+    token = os.getenv(STUDIO_TOKEN) or default_token
+    url_finder = repo_url_finder or get_studio_repo_url
+    repo_url = os.getenv(STUDIO_REPO_URL) or url_finder()
+    return token, repo_url
+
+
+def post(
+    endpoint: str,
+    token: str,
+    data: Dict[str, Any],
+    url: Optional[str] = STUDIO_URL,
+    max_retries: int = 3,
+    timeout: int = 5,
+) -> "Response":
+    endpoint = urljoin(url or STUDIO_URL, endpoint)
+    session = Session()
+    session.mount(endpoint, HTTPAdapter(max_retries=max_retries))
+
+    logger.trace("Sending %s to %s", data, endpoint)  # type: ignore[attr-defined]
+
+    headers = {"Authorization": f"token {token}"}
+    resp = session.post(endpoint, json=data, headers=headers, timeout=timeout)
+    resp.raise_for_status()
+    return resp
+
+
+def notify_refs(
+    git_remote: str,
+    default_token: Optional[str] = None,
+    studio_url: Optional[str] = None,
+    repo_url_finder: Optional[Callable[[], Optional[str]]] = None,
+    **refs: List[str],
+) -> None:
+    # TODO: Should we use git_remote to associate with Studio project
+    # instead of using `git ls-remote` on fallback?
```

Comment:
@dberenbaum, a bit of an edge case here.
What project/repository should we associate on `dvc exp push upstream`? The `upstream` remote url, or where the last `git push` was on for the branch?

Reply:
Since it's an explicit remote, I think it should be the `origin` remote url. If the user wants to use a different remote, then they shouldn't specify `origin`, right?

WRT using the upstream remote, there's some very old discussion in https://github.com/iterative/dvc/pull/6332#discussion_r679580846 and https://github.com/iterative/dvc/discussions/6427. I agree it would be nice to have some way to use the upstream remote, but I think it's not that simple since we aren't pushing a branch and it's not a blocker.

Reply:
Sorry, I meant it in terms of a Studio project. We send remote url (aka repo_url) to Studio. 
At the moment, we are sending —set-upstream url instead of the remote that is passed as argument on exp-push or exp-remove.

Reply:
Thinking about it more, it does not make sense to pass a different repo_url if you have pushed to a separate remote. I am changing it.

Reply:
Yep, makes sense, thanks @skshetry!